### PR TITLE
Solutions to discretisation exercises

### PR DESCRIPTION
**Pull request by dianetambey**
finite differences: boundary conditions and formula fix
ex 2: * missing in formula
ex 4: put in square box
sol 1: formula fix

### CODE REVIEW

File: docs/src/guide/discretisation.jl [hunk 1]
```diff
@@ -63,7 +63,7 @@ end;
 # !!! tip "Exercise 2"
 #     Show that
 #     ```math
-#        \langle e_G, e_{G'}\rangle = ∫_0^{2π} e_G(x) e_{G'}(x) d x = δ_{G, G'}
+#        \langle e_G, e_{G'}\rangle = ∫_0^{2π} e_G(x)\ast e_{G'}(x) d x = δ_{G, G'}
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:66]: Better: e_G^\ast(x)

File: docs/src/guide/discretisation.jl [hunk 2]
```diff
@@ -118,7 +118,8 @@ plot(real(ψ); label="")
 
 # Again this should match with the result above.
 #
-# **Exercise 4:** Look at the Fourier coefficients of $\psi$ ($\psi$_fourier)
+# !!! "Exercise 4"
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:121]: I think the syntax here and the next line is not correct (some keyword missing after !!! & indention in next lines

**Review by mfherbst — COMMENTED** [review]
Thanks, some nits.

File: docs/src/guide/discretisation.jl (hunk 1 above)
Comment (dianetambey) [reply docs/src/guide/discretisation.jl:66]: Thank you, I will fix that.

File: docs/src/guide/discretisation.jl (hunk 2 above)
Comment (dianetambey) [reply docs/src/guide/discretisation.jl:121]: Right thanks !

File: docs/src/guide/discretisation.jl [hunk 3]
```diff
@@ -147,7 +148,7 @@ Array(H)
 # !!! tip "Exercise 6"
 #     Increase the size of the problem, and compare the time spent
 #     by DFTK's internal diagonalization algorithms to a full diagonalization of `Array(H)`.  
-#     *Hint:* The `@time` macro is handy for this task.
+#     *Hint:* The `@benchmark` macro is handy for this task.
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:151]:
```suggestion
#     *Hint:* The `@benchmark` macro (from the [BenchmarkTools](https://github.com/JuliaCI/BenchmarkTools.jl) package) is handy for this task. Note that there are some subtleties with global variables (see the BenchmarkTools docs for details). E.g. to use it to benchmark a function like `eigen(H)` run it as
#     ```
#     using BenchmarkTools
#     @benchmark eigen($H)
#     ```
#     (note the `$`).
```

File: docs/src/guide/discretisation.jl [hunk 4]
```diff
@@ -198,22 +199,121 @@ Nrange = 10:10:100
 plot(Nrange, abs.(fconv.(Nrange) .- fconv(200)); yaxis=:log, legend=false)
 
 # ### Exercise 2
-# !!! note "TODO"
-#     This solution has not yet been written. Any help with a PR is appreciated.
+#      ```math
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:202]: No need to indent this and other math blocks.

File: docs/src/guide/discretisation.jl [hunk 5]
```diff
@@ -198,22 +199,121 @@ Nrange = 10:10:100
...
+#         \bullet \langle e_G, e_{G'}\rangle = ∫_0^{2π} e_G^\ast(x) e_{G'}(x) d x = 1/2π ∫_0^{2π} e^i(G'-G)x d x
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:203]: Why `\bullet` ?

File: docs/src/guide/discretisation.jl [hunk 6]
```diff
@@ -198,22 +199,121 @@ Nrange = 10:10:100
...
+#      ```
+# if G≠G', since the function is periodic over $[0, 2\pi]$:
+#      ```math
+#         \langle e_G, e_{G'}\rangle = \frac 1 {i2π(G'-G)}  ∫_0^{2π} (e^{ix})^{G'-G} d x = 0
+#      ```
+# if G=G':
+#     ```math
+#        \langle e_G, e_{G'}\rangle =  \frac 1 {2π} ∫_0^{2π} d x = 1
+#     ```
+# Therefore:
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:213]: Which function ? I think that's a bit wordy
```suggestion
# Since `e^{iy}` is a periodic function with period $2\pi$, $\int_0^{2\pi} e^{i m y} dy = \delta_{0,m}$. Therefore:
```

File: docs/src/guide/discretisation.jl [hunk 7]
```diff
@@ -198,22 +199,121 @@ Nrange = 10:10:100
...
+#      ```math
+#         \langle e_G, e_{G'}\rangle = δ_{G, G'}
+#      ```
+#         \bullet Assuming $V(x) = \cos(x)$:
+#     ```math
+#        \langle e_G, H e_{G'}\rangle = \frac 1 2 ∫_0^{2π} e_G^\ast(x) H e_{G'}(x) d x \left(|G|^2 \delta_{G,G'} + \delta_{G, G'+1} + \delta_{G, G'-1}\right).
+#     ```
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:220]: This is the final result, right ? I would not repeat that here. People can read it in the question.

**Review by mfherbst — COMMENTED** [review]
Some first notes.

File: docs/src/guide/discretisation.jl (hunk 5 above)
Comment (dianetambey) [reply docs/src/guide/discretisation.jl:203]: I thought I would put a bullet for each part of the equation, should I remove it ?

File: docs/src/guide/discretisation.jl [hunk 8]
```diff
@@ -198,22 +201,108 @@ Nrange = 10:10:100
...
 # ### Exercise 3
-# !!! note "TODO"
-#     This solution has not yet been written. Any help with a PR is appreciated.
+# The Hamiltonian matrix for the plane waves method can be found this way:
+## Plane waves Hamiltonian -½Δ + cos on [0, 2pi].
+
+using LinearAlgebra
+
+function build_plane_waves_matrix_cos(N::Integer)
+    # Plane wave approximation to -½Δ
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:235]: When you put in-line  comments in code segments, they need to start with ## ... Ask me later why if you are curious

File: docs/src/guide/discretisation.jl [hunk 9]
```diff
@@ -198,22 +201,108 @@ Nrange = 10:10:100
...
+    G = [float((-N + i)^2) for i in 0:2N]
+    #using results from exercice 2 for the case of cos potential, the following matrix is built for the Hamiltonian
```
Comment (mfherbst) [inline docs/src/guide/discretisation.jl:237]: Here as well: ## using ...
